### PR TITLE
Use `setup.py` directly rather than `pip install`

### PR DESCRIPTION
**Pull request by mdboom**
This results in building the project in the same directory every time, giving ccache a better chance of doing its thing.  An alternate solution to #201.

`pip install /some/local/path` has been used in asv to install the benchmarked project on the assumption that `pip` knows better about the details of how to install a Python package beyond just the usual `setup.py install`.  However, I wonder if that's really the case...

Note that I tried to use the `--build` flag for `pip install` to force it to build in a specific directory.  This seems to work for remote packages, but it's ignored when given a local path...

Attn: @pv

### CODE REVIEW

**Comment by pv** [comment]
I think pip refuses to uninstall packages installed by `setup.py install` --- at least for the case when setup.py is not using setuptools.

Perhaps just running `python setup.py build` before `pip install .` will work?
At least for Numpy, this appears to skip the build in the second step.

**Comment by pv** [comment]
The second question is then what to do with building dependencies --- if the environment in asv is not torn down every time, using ccache for those is probably not so important?

**Comment by mdboom** [comment]
> I think pip refuses to uninstall packages installed by setup.py install --- at least for the case when setup.py is not using setuptools.

> Perhaps just running python setup.py build; pip install . will work (or just build_ext instead of build)?

At least experimentally on a dummy package that uses only vanilla distutils and no setuptools, this appears to be true.  `python setup.py install` makes something that isn't `pip uninstall`-able.  `python setup.py build` followed by `pip install ./project` works, though.  That seems like a good solution, then.

As for the dependencies, you are correct that they are still being built with random paths, therefore ccache doesn't work.  But they are cached in the `env` directory tree and don't need to be built each time, so it's less of an issue.  If it were an issue, we could pass `--build ./some-static-tmp-directory` to pip when building the dependencies, since `--build` *does* work when installing from downloaded packages, just not when installing from the local filesystem.

**Comment by pv** [comment]
This works fine for me. Does it break the parallel build feature however?

**Comment by mdboom** [comment]
I *does* break the parallel build feature.  For a couple of other corner case reasons, I've been meaning to fix the parallel build feature so that each parallel thread would use its own `git clone` anyway, so this might just be onus to do that...

**Comment by mdboom** [comment]
This now makes a separate clone of the project repository for each environment (these are `--shared` clones, so their `.git` directories don't take up much space).  That way parallel building can work without clobbering another environment's build.  This also includes @pv's wheel caching support, so the last *N* builds are retained and can be reinstalled next time much faster.

The combination of parallel building, ccache working and cached wheels makes things *much* faster overall.  Too bad there isn't a tool to measure that ;)

**Comment by pv** [comment]
Seems to break asv dev, cf pv@1addad0
`asv dev` also seems to unnecessarily clone the repo.

File: asv/wheel_cache.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        for name in names[self._wheel_cache_size:]:
+            path = os.path.join(self._path, name)
+            if os.path.isdir(path):
+                shutil.rmtree(path)
+
+    def build_project_cached(self, env, package, commit_hash):
+        if self._wheel_cache_size == 0:
+            return None
+
+        if not os.path.isdir(self._path):
```
Comment (pv) [inline asv/wheel_cache.py]: This check should be removed --- the directory is first created by the call to `_get_wheel` below

**Comment by pv** [comment]
Worksforme apart from those two

**Comment by mdboom** [comment]
Thanks.  I think I've addressed those two issues.

**Comment by pv** [comment]
The isdir check probably needs to be added to `_get_wheel` pv@6762b7d ?

**Comment by mdboom** [comment]
Indeed.  I think it's fixed now.

File: asv/wheel_cache.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        if not os.path.isdir(path):
+            os.makedirs(path)
+        with open(stamp, 'wb'):
+            pass
+        return path
+
+    def _get_wheel(self, commit_hash):
+        cache_path = self._get_wheel_cache_path(commit_hash)
+
+        if not os.isdir(cache_path):
```
Comment (pv) [inline asv/wheel_cache.py]: os.path.isdir

File: asv/console.py [hunk 3]
```diff
@@ -304,6 +304,13 @@ def enable(self, verbose=False):
         else:
             self._logger.setLevel(logging.INFO)
 
+    @contextlib.contextmanager
+    def set_level(self, level):
+        orig_level = self._logger.level
+        self._logger.setLevel(level)
+        yield
+        self._logger.setLevel(orig_level)
```
Comment (pv) [inline asv/console.py]: This doesn't set the level back if an exception is raised, may be related to gh-209

Comment (pv) [inline asv/console.py]: pv@044b8d6f9c

Comment (mdboom) [inline asv/console.py]: Thanks.  I've cherry-picked 044b8d6 over to this branch.

Comment (pv) [inline asv/console.py]: I've been using this PR for the Scipy benchmarks, and didn't meet other
problems than what is fixed above.

**Pull request merged**